src/main/formats/CPS/CPS1Instr.cpp

```diff
@@ -91,4 +90,72 @@ bool CPS1SampColl::parseSampleInfo() {
     samples.push_back(sample);
   }
   return true;
-}
\ No newline at end of file
+}
+
+// ******************
+// CPS1SampleInstrSet
+// ******************
+
+CPS1OPMInstrSet::CPS1OPMInstrSet(RawFile *file,
+                               CPSFormatVer version,
+                               uint32_t offset,
+                               const std::string& name)
+    : VGMInstrSet(CPS1Format::name, file, offset, 0, name),
+      fmt_version(version) {
+}
+
+bool CPS1OPMInstrSet::parseInstrPointers() {
+  for (int i = 0; i < 128; ++i) {
+    auto offset = dwOffset + (i * sizeof(CPS1OPMInstrData));
+    if (VGMFile::readWord(offset) == 0 && VGMFile::readWord(offset+4) == 0) {
+      break;
+    }
+
+    auto instr = new CPS1OPMInstr(this, offset, sizeof(CPS1OPMInstrData), 0,
+      i, fmt::format("Instrument {}", i));
+    aInstrs.push_back(instr);
+  }
+  return true;
+}
+
+std::string CPS1OPMInstrSet::generateOPMFile() {
+  std::ostringstream output;
+  std::string header = std::string("// Converted using VGMTrans version: ") + VGMTRANS_VERSION + "\n";
+  output << header;
+
+  for (size_t i = 0; i < aInstrs.size(); ++i) {
+    CPS1OPMInstr* instr = dynamic_cast<CPS1OPMInstr*>(aInstrs[i]);
+    if (instr != nullptr) {
+      output << instr->toOPMString(i) << "\n";
+    }
```

Comment:
there's this handy c++17 construct:
```suggestion
        if (auto* instr = dynamic_cast<CPS1OPMInstr*>(aInstrs[i]); instr != nullptr) {
            output << instr->toOPMString(i) << '\n';
        }
```

Reply:
Oh right. I've used this in other places. Really glad they added this pattern to C++.